### PR TITLE
Set Jenkins node temporarily offline due to specific machine issue

### PR DESCRIPTION
**Pull request by llxia**
Currently, this feature is only enabled if params.SLACK_CHANNEL is provided.
This PR checks test build console output only if the build result is FAILURE. If the console
output contains a specific error message (i.e., There is not enough space in the file system, etc),
set the current node to be temporarily offline. If SLACK_CHANNEL is provided, send the message in
slack.

Resolve: #2370
Signed-off-by: lanxia <lan_xia@ca.ibm.com>

### CODE REVIEW

**Comment by llxia** [comment]
- if params.SLACK_CHANNEL is NOT provided, the feature will not be enabled https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/221/
- if params.SLACK_CHANNEL is set, set error to match `.*Cloning the remote Git.*` for testing purpose (Grinder/14854/):


job console output:
```
Exception: hudson.AbortException: script returned exit code 128
[Pipeline] timeout
Timeout set to expire in 5 min 0 sec
[Pipeline] {
[Pipeline] echo
Grinder #14854 result is FAILURE. Checking console log for specific errors...
[Pipeline] echo
Grinder #14854 failed on ub20x64rt3-1 due to .*Cloning the remote Git.*: https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/14854/ 
Set ub20x64rt3-1 temporarily offline! https://hyc-runtimes-jenkins.swg-devops.com/computer/ub20x64rt3-1
[Pipeline] slackSend
Slack Send Pipeline step running, values are - baseUrl: <empty>, teamDomain: ibm-cloud, channel: #rt-jenkins, color: <empty>, botUser: false, tokenCredentialId: 86f8f7c2-0c7a-4f7f-a153-cae6265a3f91, notifyCommitters: false, iconEmoji: <empty>, username: <empty>, timestamp: <empty>
[Pipeline] }
[Pipeline] // timeout
[Pipeline] }
[Pipeline] // node
[Pipeline] }
[Pipeline] // stage
[Pipeline] End of Pipeline
Finished: FAILURE
````
Slack:


Jenkins node:

**Comment by llxia** [comment]
FYI @pshipton @AdamBrousseau

**Comment by AdamBrousseau** [comment]
I think this feature hinging on SLACK_CHANNEL parameter is a bit non-intuitive.
I am also a bit concerned about the feature itself but would be interested in putting it through a trial period. We may want to move the error list rules to a separate config file.

**Comment by llxia** [comment]
> I think this feature hinging on SLACK_CHANNEL parameter is a bit non-intuitive.

I agree. The reason for using `SLACK_CHANNEL` is that I do not want to introduce another parameter and I do not want to enable it for all Jenkins servers. If this feature works well in the trial run, we may not need any flag (this can be set as the default behavior)

> We may want to move the error list rules to a separate config file.

yes, we discussed this in the AQA meeting today. We could have a separate config file or user-specified file for different purposes (i.e., different Jenkins, different platforms, etc). And we would like to add further actions, such as 
- ob failed with no space -> trigger clean job on the node
- job failed with git issue due to installation -> open infra issue/rerun playbook
- retrigger the test job as the failure is due to the machine issue

This PR is the first version. We can enhance it later if it works as expected.

**Comment by llxia** [comment]
Updated to resolve merge conflicts.

**Review by smlambert — COMMENTED** [review]
LGTM @llxia, I would like to try it at the Adopt Jenkins server.  As long as the Jenkins server is configured to a particular Slack workspace and you set SLACK_CHANNEL to be a channel in that workspace, is that all that is required?

**Review by smlambert — APPROVED** [review]

**Comment by llxia** [comment]
> As long as the Jenkins server is configured to a particular Slack workspace and you set SLACK_CHANNEL to be a channel in that workspace, is that all that is required?

yes, that's correct.

**Comment by smlambert** [comment]
@sophia-guo - waiting on your review.

File: buildenv/jenkins/openjdk_tests [hunk 1]
```diff
@@ -239,42 +239,93 @@ if (JDK_VERSIONS.size() > 1 || JDK_IMPLS.size() > 1 || PLATFORMS.size() >1 || PL
...
+                            userRemoteConfigs: [[url: "${gitConfig.getUrl()}"]]
+                        ]
+                    }
+                    jenkinsfile = load "${WORKSPACE}/openjdk-tests/buildenv/jenkins/JenkinsfileBase"
+                    jenkinsfile.testBuild()
+                } catch (Exception e) {
+                    println("Exception: " + e.toString())
+                    // build result may not be updated correctly at the moment (see https://issues.jenkins.io/browse/JENKINS-56402)
+                    // if there is an exception and build result is not FAILURE, set currentBuild.result to FAILURE
+                    if (currentBuild.currentResult != 'FAILURE' && currentBuild.result != 'FAILURE') {
```
Comment (sophia-guo) [inline buildenv/jenkins/openjdk_tests:280]: Looks like this if condition is to work with the following if condition https://github.com/adoptium/aqa-tests/pull/2530/files#diff-ebf20f64c486ae48bd60b965c555076519f9d5623789f4574467d3051bb530f7R286. 

If the effort is to cover the case result == null probably would be better to try isWorseThan(SUCCESS) https://javadoc.jenkins-ci.org/hudson/model/Result.html#isWorseThan-hudson.model.Result-

Comment (sophia-guo) [reply buildenv/jenkins/openjdk_tests:280]: Yeah, understand that the problem is we can not rely on currentBuild.currentResult or currentBuild.result on the finally clause for notifications or other things as  they aren't set until we get out of the finally. So we need to set currentBuild.result manually. We can set the result if there is exception. My suggestion is could we do the following:

```
...
  } catch(e) {
    currentBuild.result = 'FAILURE'
  } finally {
   ...
    if (currentBuild.result == 'FAILURE')  {// or currentBuild.current == 'FAILURE')  both works
        code for error checking
     }
    ...
  }

Comment (llxia) [reply buildenv/jenkins/openjdk_tests:280]: We do not need to override the result if it is `FAILURE `

Comment (llxia) [reply buildenv/jenkins/openjdk_tests]: Talked with @sophia-guo , `currentResult` is set to `SUCCESS` if `result` is null. In other cases, `currentResult` is set to `result`. https://github.com/jenkinsci/workflow-support-plugin/pull/31/files

PR is updated.

**Comment by smlambert** [comment]
Noting https://github.com/eclipse-openj9/openj9/issues/12637#issuecomment-841755027

Does the error reported there mean that git is not available on the machine?  If so, it would be a candidate error for marking a machine offline.

```
The recommended git tool is: git
[Pipeline] }
[Pipeline] // node
[Pipeline] }
[Pipeline] // stage
[Pipeline] End of Pipeline
java.io.EOFException
	at java.base/java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:3001)
	at java.base/java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3496)
	at java.base/java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:1058)
	at java.base/java.io.ObjectInputStream.<init>(ObjectInputStream.java:415)
	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:49)
	at hudson.remoting.Command.readFrom(Command.java:142)
	at hudson.remoting.Command.readFrom(Command.java:128)
	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)
	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:63)
Caused: java.io.IOException: Unexpected termination of the channel
	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:77)
```

**Comment by pshipton** [comment]
I think it's a temporary network problem. For everything in https://github.com/eclipse-openj9/openj9/issues/12637 the machines are fine (in general) and run subsequent jobs successfully.

**Review by sophia-guo — APPROVED** [review]
LGTM

**Comment by AdamBrousseau** [comment]
I think the `readFully` error is the slave.jar crashing, maybe OOM. It seems to be a windows thing. Just my anecdotal experience though. I can't recall at the moment why I think that. The git line is standard output.

**Comment by smlambert** [comment]
@llxia - can you resolve the conflicts on this one please?

**Comment by llxia** [comment]
I am currently working on dynamic VM agent feature. It needs to update the same block in the same file. It may be better to rebase this PR once we have the dynamic VM agent feature up and running to avoid future merging conflicts.

**Comment by llxia** [comment]
The conflicts are resolved.

Retested: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/913/
Grinder/1627

**Comment by llxia** [comment]
Could I get this PR reviewed again? The conflicts are resolved.

**Review by smlambert — APPROVED** [review]
LGTM, if you are not waiting on anything else, I am happy to merge it.

**Pull request merged**